#118: Ability to handle parameters in GatewayClass

sergioifg94 opened the pull request:
## Description 

> Closes #118 

Ability to get parameters from a resourced referenced in the `parametersRef` of the GatewayClass

## Verification steps

As the parameters are currently not used, in order to verify this functionality, ensure that the adequate status condition is reconciled when the `parametersRef` is misconfigured.

1. Start the local development clusters
    ```sh
    MCTC_WORKLOAD_CLUSTERS_COUNT=1 make local-setup
    ```

3. Run the controller
    ```sh
    (export $(cat ./controller-config.env | xargs) && export $(cat ./aws-credentials.env | xargs) && make build-controller install run-controller)
    ```

1. Create the parameters ConfigMap
    ```sh
    kubectl create configmap gateway-params --from-file=params=config/samples/gatewayclass_params.json -n multi-cluster-traffic-controller-system
    ```

5. Create the GatewayClass
    ```yaml
    apiVersion: gateway.networking.k8s.io/v1beta1
    kind: GatewayClass
    metadata:
      name: mctc-gw-istio-external-instance-per-cluster
    spec:
      controllerName: "kuadrant.io/mctc-gw-controller"
      parametersRef:
        group: ""
        kind: "ConfigMap"
        namespace: "multi-cluster-traffic-controller-system"
        name: "gateway-params"
    ```

7. Create the Gateway with `.spec.gatewayClassName = mctc-gw-istio-external-instance-per-cluster`
    ```sh
    kubectl create namespace mctc-tenant
    kubectl apply -n mctc-tenant -f gateway.yaml
    ```

1. Verify that once reconciled, the annotation is set
    ```sh
    kubectl get gateway/example-gateway -n mctc-tenant
    ```
    ```yaml
    metadata:
      name: example-gateway
      annotations:
        mctc-syncer-patch/kind-mctc-workload-1: [{"op":"replace","path":"/spec/gatewayClassName","value":"istio"}] 
    ```

makslion commented:
Verified GatewayClass gets `InvalidParameters` field 
![image](https://user-images.githubusercontent.com/37799801/227180792-40648190-c780-469c-8ccf-fbf9ec7c63b1.png)

makslion commented on pkg/controllers/gateway/gateway_controller.go:
```diff
@@ -117,9 +117,32 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, nil
 	}
 
+	// TODO: Use return value
```
Aren't we using the return value at line 145? :thinking:

makslion commented on pkg/controllers/gateway/gateway_controller.go:
```diff
@@ -149,7 +172,7 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 // Configures Gateway tls & dns for each cluster it targets.
 // Returns the programmed status, a list of clusters that were programmed, if the gateway should be requeued, and any error
-func (r *GatewayReconciler) reconcileGateway(ctx context.Context, previous gatewayv1beta1.Gateway, gateway *gatewayv1beta1.Gateway) (metav1.ConditionStatus, []string, bool, error) {
+func (r *GatewayReconciler) reconcileGateway(ctx context.Context, previous gatewayv1beta1.Gateway, gateway *gatewayv1beta1.Gateway, params *Params) (metav1.ConditionStatus, []string, bool, error) {
```
_ *Params?

sergioifg94 replied:
I think that given that we intend to use that parameter in the future it makes sense to leave it named. I'd use the `_` when it's a parameter that we don't care about (we don't yet, but we will)

makslion commented on pkg/controllers/gateway/params.go:
```diff
@@ -0,0 +1,89 @@
+
+type ParamsResolver func(context.Context, client.Client, gatewayv1beta1.ParametersReference) (*Params, error)
+
+var paramsResolvers map[schema.GroupKind]ParamsResolver = map[schema.GroupKind]ParamsResolver{}
```
the type here could be omitted as well, right?

maleck13 commented on pkg/controllers/gateway/gateway_controller.go:
```diff
@@ -117,9 +117,32 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+
+	// If the GatewayClass parametrers are invalid, update the status
+	// and stop reconciling
+	if IsInvalidParamsError(err) {
```
Nit: This might read better as 

``` 
if err !=nil && IsInvalidParamsError(err) {
...

}

if err != nil{
   return ctrl.Result{}, nil
}
```

Why do we ignore the err on line 122 should that be returned?

sergioifg94 replied:
Makes sense, I'll make those changes. About the err being ignored, I missed it, it should indeed be returned

maleck13 commented on pkg/controllers/gateway/params.go:
```diff
@@ -0,0 +1,89 @@
+		Group: corev1.GroupName,
+		Kind:  "ConfigMap",
+	}
+	paramsResolvers[configMapGroupKind] = fromConfigMap
```
I wonder if we need this, I think we are only ever likely to have one params object type

sergioifg94 replied:
The motivation for it is to be able to support multiple types in case we want to for example define a CRD for the parameters. So the user can have the option to define the parameters in a ConfigMap or in a CR and the controller will be able to handle it

maleck13 commented on pkg/controllers/gateway/params.go:
```diff
@@ -0,0 +1,89 @@
+		return nil, &InvalidParamsError{fmt.Sprintf("unable to retrieve parameters for GroupKind %s", groupKind.String())}
+	}
+
+	params, err := resolveParams(ctx, client, *gatewayClass.Spec.ParametersRef)
```
could this be simplified to

```
return resolveParams()
```

maleck13 commented on pkg/controllers/gateway/params.go:
```diff
@@ -0,0 +1,89 @@
+	return e.message
+}
+
+func IsInvalidParamsError(err error) (is bool) {
```
I always wonder if the default for these should be

```
if err == nil{
return false
}
```

sergioifg94 replied:
That's an interesting question, because it would be possible to define
```go
var err1 *InvalidParamsError = nil
var err2 error = nil
```

Both err1 and err2 would be `nil` but passing err1 to `IsInvalidParamsError` would return true, and passing err2 would return false.
I think it makes sense to leave it like this and leave the nil check responsibility to the caller of `IsInvalidParamsError`

maleck13 reviewed (commented):
We have a valid use case for the parameters in adding in the supported data plane gateway controller to use (istio)
So we could actually use the code here and create a default configmap params with the gateway controller to use. This param is set as a transform on the gateway resource

maleck13 commented:
couple of code comments and a suggested change to make the code do some actual work for us

sergioifg94 commented:
Updated the PR with the suggestions, added some unit tests and updated the verification steps to verify that the patch annotation is set from the value specified in the parameters

maleck13 commented:
@sergioifg94 just one question before merge. Prob worth having this params created as part of local set up so it is there and being used?

sergioifg94 commented:
@maleck13 I added a commit to create the params in the local-setup and the `parametersRef` section filled by default in the sample gateway class

maleck13 commented on hack/local-setup.sh:
```diff
@@ -224,6 +224,17 @@ initController() {
+
+  echo "Adding ${clusterName} to gatewayclass params"
+
+  ${YQ_BIN} -i -P -o=json '.downstreamClasses["'$clusterName'"] = "istio"' config/samples/gatewayclass_params.json
```
@sergioifg94 why are we keying it off the cluster name? The idea with the gatewayclass params is it is configuration to be applied to gateways of this class, so I would expect all gateways of this class would have the same downstream class

maleck13 commented on pkg/controllers/gateway/gateway_controller.go:
```diff
@@ -164,6 +188,15 @@ func (r *GatewayReconciler) reconcileGateway(ctx context.Context, previous gatew
 	managedHosts := []ManagedHost{}
 
+	// Set the annotations to sync the class name from the parameters
+	for cluster, downstreamClass := range params.GetDownstreamClasses() {
```
yeah same here, I think for the downstream class, each instance would have the same class

maleck13 reviewed (commented):
I don't think the params should be keyed of the clustername, the idea with the params is it is additional config for a given gateway class and so generic

maleck13 commented:
@mikenairn I see your PR depends on this one. We can probably merge this, I would just prefer the params weren't keyed against a cluster. The gatewayclass params, IMO are intended as class level configuration rather than cluster level configuration cc @sergioifg94

sergioifg94 commented:
I was planning on waiting until patching annotations for all clusters is supported, but we can merge this as is for now and make those changes later

sergioifg94 commented:
@maleck13 I just updated the PR to follow your suggestion, so the parameter is now `downstreamClass` and the patch is applied to all clusters
